File: lib/policy.ts
export type GateInputs = {
  deploymentId: string
  riskScore: number
  driftScore: number
  signatureValid: boolean
  attestationValid: boolean
  regressionPassed: boolean
  approverOneId: string
  approverOneApproved: boolean
  approverTwoId: string
  approverTwoApproved: boolean
}

export type GateDecision = {
  status: "approved" | "manual_hold"
  reasons: string[]
  policyVersion: string
  evaluatedAt: string
}

export const POLICY_VERSION = "edge-governance-v1.0"

export function evaluateGate(
  input: GateInputs,
  evaluatedAt = new Date().toISOString()
): GateDecision {
  const reasons: string[] = []
  const approverOneId = input.approverOneId.trim()
  const approverTwoId = input.approverTwoId.trim()

  if (!input.deploymentId.trim()) reasons.push("Deployment identity is missing")
  if (!approverOneId || !approverTwoId) reasons.push("Both approver identities are required")
  if (approverOneId && approverOneId === approverTwoId) {
    reasons.push("Approvers must be two distinct identities")
  }
  if (!input.approverOneApproved || !input.approverTwoApproved) {
    reasons.push("Two-person approval is incomplete")
  }
  if (input.riskScore > 30) reasons.push("Risk score exceeds the policy threshold of 30")
  if (input.driftScore > 25) reasons.push("Drift score exceeds the policy threshold of 25")
  if (!input.signatureValid) reasons.push("Bundle signature is invalid")
  if (!input.attestationValid) reasons.push("Device attestation is invalid")
  if (!input.regressionPassed) reasons.push("Qualification regression gate failed")

  return {
    status: reasons.length === 0 ? "approved" : "manual_hold",
    reasons,
    policyVersion: POLICY_VERSION,
    evaluatedAt,
  }
}

File: demos/streamlit/policy.py
from dataclasses import dataclass
from datetime import datetime, timezone

POLICY_VERSION = "edge-governance-v1.0"
RISK_THRESHOLD = 30
DRIFT_THRESHOLD = 25


@dataclass(frozen=True)
class GateInput:
    deployment_id: str
    risk_score: float
    drift_score: float
    signature_valid: bool
    attestation_valid: bool
    regression_passed: bool
    approver_one_id: str
    approver_one_approved: bool
    approver_two_id: str
    approver_two_approved: bool


def evaluate_gate(value: GateInput) -> dict:
    reasons: list[str] = []
    approver_one = value.approver_one_id.strip()
    approver_two = value.approver_two_id.strip()

    if not value.deployment_id.strip():
        reasons.append("Deployment identity is missing")
    if not approver_one or not approver_two:
        reasons.append("Both approver identities are required")
    if approver_one and approver_one == approver_two:
        reasons.append("Approvers must be two distinct identities")
    if not value.approver_one_approved or not value.approver_two_approved:
        reasons.append("Two-person approval is incomplete")
    if value.risk_score > RISK_THRESHOLD:
        reasons.append(f"Risk score exceeds the policy threshold of {RISK_THRESHOLD}")
    if value.drift_score > DRIFT_THRESHOLD:
        reasons.append(f"Drift score exceeds the policy threshold of {DRIFT_THRESHOLD}")
    if not value.signature_valid:
        reasons.append("Bundle signature is invalid")
    if not value.attestation_valid:
        reasons.append("Device attestation is invalid")
    if not value.regression_passed:
        reasons.append("Qualification regression gate failed")

    return {
        "status": "approved" if not reasons else "manual_hold",
        "reasons": reasons,
        "policy_version": POLICY_VERSION,
        "evaluated_at": datetime.now(timezone.utc).isoformat(),
    }
